File: android/app/src/main/java/com/lilru/liftr/domain/StrengthSetMultiplicity.kt
package com.lilru.liftr.domain

import kotlin.math.max

/**
 * Paridad con [strengthSetMultiplicities] en Liftr/StrengthSetMultiplicity.swift.
 */
fun strengthSetMultiplicities(sortedSetNumbers: List<Int>): List<Int> {
    val r = sortedSetNumbers.size
    if (r <= 0) return emptyList()
    if (sortedSetNumbers == (1..r).toList()) {
        return List(r) { 1 }
    }
    return sortedSetNumbers.map { max(it, 1) }
}

data class StrengthSetRowWithMultiplier<T>(
    val row: T,
    val multiplier: Int,
    val lineOrdinal: Int
)

fun <T> strengthSetRowsWithMultiplicities(
    rows: List<T>,
    orderIndex: (T) -> Int?,
    id: (T) -> Int,
    setNumber: (T) -> Int
): List<StrengthSetRowWithMultiplier<T>> {
    val sorted = rows.sortedWith(
        compareBy<T> { orderIndex(it) ?: Int.MAX_VALUE }.thenBy { id(it) }
    )
    val mults = strengthSetMultiplicities(sorted.map(setNumber))
    return sorted.mapIndexed { index, row ->
        StrengthSetRowWithMultiplier(
            row = row,
            multiplier = mults.getOrElse(index) { 1 },
            lineOrdinal = orderIndex(row) ?: (index + 1)
        )
    }
}

fun strengthSetSummaryWithMultiplier(body: String, multiplier: Int): String =
    if (multiplier > 1) "$multiplier× · $body" else body

data class StrengthWeightSegmentWire(
    val reps: Int,
    val weightKg: Double
)

fun exerciseSetSegmentVolumeKg(
    reps: Int?,
    weightKg: Double?,
    weightSegments: List<StrengthWeightSegmentWire>?
): Double {
    if (!weightSegments.isNullOrEmpty() && weightSegments.size >= 2) {
        return weightSegments.sumOf { max(it.reps, 0).toDouble() * max(it.weightKg, 0.0) }
    }
    val repVal = max(reps ?: 0, 0)
    val weightVal = weightKg ?: 0.0
    return repVal.toDouble() * max(weightVal, 0.0)
}

fun <T> strengthDetailVolumeKg(
    setsByExercise: Map<Int, List<T>>,
    orderIndex: (T) -> Int?,
    id: (T) -> Int,
    setNumber: (T) -> Int,
    reps: (T) -> Int?,
    weightKg: (T) -> Double?,
    weightSegments: (T) -> List<StrengthWeightSegmentWire>?
): Double {
    var total = 0.0
    for (rows in setsByExercise.values) {
        val paired = strengthSetRowsWithMultiplicities(rows, orderIndex, id, setNumber)
        for (item in paired) {
            val segmentVolume = exerciseSetSegmentVolumeKg(
                reps = reps(item.row),
                weightKg = weightKg(item.row),
                weightSegments = weightSegments(item.row)
            )
            total += segmentVolume * item.multiplier
        }
    }
    return total
}

File: Liftr/StrengthSetMultiplicity.swift
import Foundation

func strengthSetMultiplicities(sortedSetNumbers: [Int]) -> [Int] {
    let r = sortedSetNumbers.count
    guard r > 0 else { return [] }
    if sortedSetNumbers == Array(1...r) {
        return Array(repeating: 1, count: r)
    }
    return sortedSetNumbers.map { max($0, 1) }
}

func strengthSortedSetRows<Row>(
    _ rows: [Row],
    orderIndex: (Row) -> Int?,
    id: (Row) -> Int,
    setNumber: (Row) -> Int
) -> [Row] {
    rows.sorted { a, b in
        let ao = orderIndex(a) ?? Int.max
        let bo = orderIndex(b) ?? Int.max
        if ao != bo { return ao < bo }
        return id(a) < id(b)
    }
}

func strengthSetRowsWithMultiplicities<Row>(
    _ rows: [Row],
    orderIndex: (Row) -> Int?,
    id: (Row) -> Int,
    setNumber: (Row) -> Int
) -> [(row: Row, multiplier: Int)] {
    let sorted = strengthSortedSetRows(rows, orderIndex: orderIndex, id: id, setNumber: setNumber)
    let mults = strengthSetMultiplicities(sortedSetNumbers: sorted.map(setNumber))
    return zip(sorted, mults).map { ($0, $1) }
}

struct StrengthDetailAggregates {
    let totalSets: Int
    let totalReps: Int
    let avgRpe: Double?
}

func exerciseSetSegmentVolumeKg(
    reps: Int?,
    weightKg: Decimal?,
    weightSegments: [StrengthWeightSegWire]?
) -> Double {
    if let segments = weightSegments, segments.count >= 2 {
        return segments.reduce(0.0) { partial, segment in
            partial + Double(max(segment.reps, 0)) * max(segment.weight_kg, 0)
        }
    }
    let repVal = max(reps ?? 0, 0)
    let weightVal = weightKg.map { NSDecimalNumber(decimal: $0).doubleValue } ?? 0
    return Double(repVal) * max(weightVal, 0)
}

func strengthDetailVolumeKg<Row>(
    setsByExercise: [Int: [Row]],
    orderIndex: (Row) -> Int?,
    id: (Row) -> Int,
    setNumber: (Row) -> Int,
    reps: (Row) -> Int?,
    weightKg: (Row) -> Decimal?,
    weightSegments: (Row) -> [StrengthWeightSegWire]?
) -> Double {
    var total = 0.0
    for rows in setsByExercise.values {
        let paired = strengthSetRowsWithMultiplicities(
            rows,
            orderIndex: orderIndex,
            id: id,
            setNumber: setNumber
        )
        for (row, mult) in paired {
            let segmentVolume = exerciseSetSegmentVolumeKg(
                reps: reps(row),
                weightKg: weightKg(row),
                weightSegments: weightSegments(row)
            )
            total += segmentVolume * Double(mult)
        }
    }
    return total
}

func strengthDetailAggregates<Row>(
    setsByExercise: [Int: [Row]],
    orderIndex: (Row) -> Int?,
    id: (Row) -> Int,
    setNumber: (Row) -> Int,
    reps: (Row) -> Int?,
    rpe: (Row) -> Decimal?
) -> StrengthDetailAggregates {
    var totalSets = 0
    var totalReps = 0
    var rpeWeightedSum = 0.0
    var rpeWeightTotal = 0.0

    for rows in setsByExercise.values {
        let paired = strengthSetRowsWithMultiplicities(
            rows,
            orderIndex: orderIndex,
            id: id,
            setNumber: setNumber
        )
        for (row, mult) in paired {
            totalSets += mult
            let repVal = max(reps(row) ?? 0, 0)
            totalReps += repVal * mult
            if let rpeDec = rpe(row) {
                let r = NSDecimalNumber(decimal: rpeDec).doubleValue
                let m = Double(mult)
                rpeWeightedSum += r * m
                rpeWeightTotal += m
            }
        }
    }

    let avgRpe = rpeWeightTotal > 0 ? (rpeWeightedSum / rpeWeightTotal) : nil
    return StrengthDetailAggregates(
        totalSets: totalSets,
        totalReps: totalReps,
        avgRpe: avgRpe
    )
}
